### DISCUSSION TITLE
Credentials Provider: Throw custom error message or status code from authorize() instead of only "CallbackRouteError"

### DISCUSSION THREAD

**Discussion by shgrth** [nextauthjs/next-auth#8999] [Ideas]
### Goals

1. Throw custom error message or status code from authorize() instead of "CallbackRouteError" and status 200.

2. Sign In process (when credentials are correct) sometimes errors with "User not confirmed" in my backend and I would like to return a custom error message OR at least a status code to the front-end so I can render a UI to enter confirmation code. Sign In is called from next-auth/react with redirect:false.

### Background

Our company's online platform has been using Next-Auth v4 with Credentials Provider and a custom authorize() function that initiates an AWS Cognito userauth and returns the tokens for jwt callback to handle and organize the data.

Our Sign In page will automatically render a UI for the user to enter confirmation code once they attempt to sign in and Cognito returns an error stating "User not confirmed.". We catch this error within authorize() and throw it to the front-end so our app can render the respective UI.

However, now with v5, the ONLY error message I get to the front end says "CallbackRouteError" with a status code of 200 OK. There is no way even to throw a custom error status code to the front end.

### Proposal

This is specific to Next-Auth/Authjs v5.

In Credentials Provider, please allow us to throw a custom error message or a status code in the return value from the signIn function when redirect is set to false. 

We require custom errors like "User not confirmed", "Invalid credentials", or "User suspended.". 

Or let us throw integer error codes in the status and render a UI on the front-end based on these codes.

**Comment by luisjuniorj** [comment]
Use the 5.0.0-beta.3 version is working there.

**Reply by AhmedBaset** [reply]
I'm using 5.0.0-beta.3 It doesn't send the custom error to the client

**Comment by Pe6ko13** [comment]
up

**Reply by DevYemi** [reply]
Please is there any new update on this ??.

Throwing an error like this used to work in v4:
````
 CredentialsProvider({
            async authorize(credentials: any) {
                let urlencoded = new URLSearchParams();

                urlencoded.append('email', credentials.email);
                urlencoded.append('password', credentials.password);
                const res = await AuthApiInstance.signIn<UserSchemaType>(urlencoded);

                console.log(res);

                if (res.isError && !res.resData) {
                    throw new Error(JSON.stringify({ errors: res.error, status: res.statusCode }))
                };

                if (res.resData?.data.token) {
                    encryptCookieDataAction(res.resData?.data.token, EncryptionCookiesKeys.AdminToken)
                }

                return res.resData?.data as any
            },
            credentials: {},
        })
````

Now I am only getting "CallbackRouteError", any help will be appreciated

**Reply by localhost-chetan** [reply]
Same here, I'm also frustrated with this error
And now i'm just returning either the User object or null and can can only display a single error message in form which is "Invalid credentials"

  async authorize(credentials): Promise<User | null> {
            const validatedData = formSchema.safeParse(credentials)

            if (validatedData.success) {
                const { email, password } = validatedData.data

                const user = await getUserByEmail(email)

                if (!!user) {
                    const isPasswordMatch = await compare(password, user.hashedPassword)
                    if (isPasswordMatch) {
                        // It indicates all credentials are valid
                        return user;
                    } else {
                        // throw new Error(`Incorrect password`)
                        return null;
                    }
                } else {
                    return null;
                    // throw new Error(`User not found`)
                }
            } else {
                return null
                // throw new Error(`Invalid credentials`)
            }
        }

**Comment by Pe6ko13** [comment]
Sorry, no updates yet,

I'm getting only "CallbackRouteError" as well

Greetings Petar

**Comment by DavidCodina** [comment]
Same issue here with 5.0.0-beta.5. 

One possible workaround would be to use a server action to handle error logic in conjunction with the server-side signIn(). However, there seems to be an issue with the server-side signIn() not syncing with the client-side SessionProvider/useSession() such that user logs out status will be 'unauthenticated'. New user signs in, status still is authenticated (until page refresh).

**Comment by Greek** [comment]
I pushed out a PR for this. See #9801

**Reply by DavidCodina** [reply]
See issue [9504](https://github.com/nextauthjs/next-auth/issues/9504).  It's kind of related. Anyways, I realized that there's a way to send back custom error messaging. Again, it requires using both a `login()` server action and the SERVER SIDE `signIn()`. However, this requires some extra steps in order to keep the client side `SessionProvider` in sync. First create a `login()` server action:

```
import bcrypt from 'bcryptjs'
import prisma from 'db'
import { signIn } from 'auth'

type LoginResponse = {
  data: any
  message: string
  success: boolean
}

/* ========================================================================

======================================================================== */

export const login = async ({ email, password }: { email: string; password: string; }): Promise<LoginResponse> => {
  if (!email || !password) {
    return {
      data: null,
      message: 'Email and/or password is missing.',
      success: false
    }
  }

  try {
    const existingUser = await prisma.user.findUnique({
      where: { email: email as string }
    })

    if (!existingUser || !existingUser.email || !existingUser.password) {
      return {
        data: null,
        message: 'Invalid credentials.',
        success: false
      }
    }

    const isMatch = await bcrypt.compare(
      password as string,
      existingUser.password as string
    )

    if (!isMatch) {
      return {
        data: null,
        message: 'Invalid credentials.',
        success: false
      }
    }

    await signIn('credentials', {
      email,
      password,
      redirect: false  // Prevent Next.js from throwing Redirect error.
    })

    return {
      data: null,
      message: 'User credentials are valid.',
      success: true
    }
  } catch (err: unknown) {
    if (err instanceof Error) {
      console.log({ name: err.name, message: err.message })
    }
    return {
      data: null,
      message: 'Server error.',
      success: false
    }
  }
}

```
Then create a custom `SessionProvider` that is wrapped in a context. This will allow you to create, use and expose `fetchSession()`.

```
'use client '

import {
  ReactNode,
  useEffect,
  createContext,
  useContext,
  useState,
  useCallback
} from 'react'

import { Session } from 'next-auth'
import {
  SessionProvider as NextSessionProvider,
  getSession
} from 'next-auth/react'

type SessionProviderProps = {
  children: ReactNode
}

export type MaybeSession = Session | null

export type SessionContextValue = {
  fetchSession: () => void
}

const SessionContext = createContext({} as SessionContextValue)

/* ========================================================================

======================================================================== */

export const SessionProvider = ({ children }: SessionProviderProps) => {
  const [session, setSession] = useState<MaybeSession>(null)

  const fetchSession = useCallback(async () => {
    const session = await getSession()
    setSession(session)
  }, [])

  useEffect(() => {
    fetchSession()
  }, [fetchSession])

  return (
    <SessionContext.Provider
      value={{
        fetchSession
      }}
    >
      <NextSessionProvider session={session}>{children}</NextSessionProvider>
    </SessionContext.Provider>
  )
}

export function useSessionContext() {
  const value = useContext(SessionContext)
  return value
}
```

Then in your `LoginForm`'s submit function do something like this:

```
const handleCredentialsLogin = async (e: any) => {
    e.preventDefault()
    setLoading(true)

    login({ email, password })
      .then((res) => {
        if (res.success === false) {
          toast.error('Login unsuccessful.', { autoClose: false })
          return res
        }

        fetchSession()
        toast.success('Login success.')
        router.replace(DEFAULT_LOGIN_REDIRECT) // i.e., '/settings', '/user', etc.
        return res
      })
      .catch((_err) => {
        toast.error('Unable to log in.', { autoClose: false })
      })
      .finally(() => {
        setLoading(false)
        setEmail('')
        setPassword('')
      })
  }
  
```
The important part in both the custom `SessionProvider` and the submit function is the `fetchSession()` function.  This is what will keep the session in sync between client and server. 
  
Going back to the original point, the `login()` server action will allow you to make checks against the provided credentials and respond back with meaningful error messaging (since Credentials `authorize()` seems to no longer support that). Note also that it's still important to do all of the same checks in the `authorize()` function since a user could just hit  the "http://localhost:3000/api/auth/callback/credentials" directly.
 
Admittedly, the current state of v5 beta is painfully unintuitive. Hope this helps.

**Reply by Greek** [reply]
I ended up just creating my own error class and assigning the `type` property to the message thru the constructor. Works just as expected ;)

```ts
/* eslint-disable @typescript-eslint/no-explicit-any */
import { AuthError } from "next-auth";

export class CustomAuthError extends AuthError {
  static type: string;

  constructor(message?: any) {
    super();

    this.type = message;
  }
}
```

The probably very right and best way to do this right is to create a class with your custom `type` and handle it on the client:

```ts
import { AuthError } from "next-auth";

export class InvalidEmailPasswordError extends AuthError {
  static type = "InvalidEmailPassword"
}
```

**Reply by riyoneri** [reply]
Hello,
Could you provide the way you used this CustomAuthError class cause it is giving me headache, For me I want to carry object of errors in my custom class, so that i can display validation errors on each inputs accordingly

how did you use it, or how can i do my custom error class so that i carry object not the string

**Reply by DavidCodina** [reply]
I have a file in src/CustomAuthErrors where I have these two:

```
import { CredentialsSignin } from 'next-auth'

export class InvalidLoginError extends CredentialsSignin {
  // This can be changed, but only to another ErrorType (e.g., 'AccessDenied')
  // static type = "CredentialsSignin"
  code = 'invalid_credentials'
}

export class UnverifiedEmailError extends CredentialsSignin {
  code = 'email_unverified'
}
```

Then in the authorize function I do this:

```
async authorize({ email, password }) {
  if (!email || !password) {
    throw new InvalidLoginError()
  }
  // ...
}
```
You could create a class that accepts an object as an argument, and then runs a switch statement to set the code. Input field validation should run on the client BEFORE executing authentication. That said, it's also a best practice to be opaque in regard to the actual reasons for authentication fails (i.e., 'invalid_credentials' ).

**Comment by balazsorban44** [comment]
Hi all! I want to support custom errors, please check out my proposal at https://github.com/nextauthjs/next-auth/issues/9099#issuecomment-1913754338 and leave a comment!

Note. If you want to agree/disagree only, just :+1: or :-1:. Only comment, if you have a meaningful suggestion to keep it effective! Thanks!

**Reply by sarowarhosen03** [reply]
agree with you

**Comment by Flavien-Pensato** [comment]
Working with "next-auth": "4.24.4" and it's already working with the Error class.

``
throw new Error("CUSTOM_MESSAGE")
``

**Reply by yogiprsetya** [reply]
How about response status?

**Reply by Flavien-Pensato** [reply]
> How about response status?

It's a server action. Their is no status response

**Comment by kazemmdev** [comment]
I was facing a problem where signIn wasn't working as expected in login forms with version `5.0.0-beta.16`. To achieve a custom user login experience, I implemented a server action function as an alternative approach. This function replaces the direct use of signIn within the login form.

```ts
"use server"

import { DEFAULT_LOGIN_REDIRECT } from "@/routes"
import { AuthError } from "next-auth"

import { signIn } from "@/lib/auth"

export const login = async (phone: string, code: string, callbackUrl?: string | null) => {
  try {
    await signIn("phone", { phone, code, redirectTo: callbackUrl || DEFAULT_LOGIN_REDIRECT })
  } catch (error) {
    console.error("phone login: ", error)
    if (error instanceof AuthError) {
      return { error: "error", message: error.message, status: 401 }
    }

    throw error
  }
}

```

**Comment by maksof-sarwar** [comment]
just return Promise.reject(new Error(" CUISTOMER MESSAGE "));

**Comment by dsmabulage** [comment]
If we throw a error inside authorize function how can we show it in the client side in the credential provider

**Reply by mksglu** [reply]
Same question from my side and have not any solutions

**Reply by Greek** [reply]
Check out #9099. If that doesn’t work try out one of the workarounds I posted [here](https://github.com/nextauthjs/next-auth/discussions/8999#discussioncomment-8250545).

**Reply by dsmabulage** [reply]
> Check out #9099. If that doesn’t work, try out one of the workarounds I posted [here](https://github.com/nextauthjs/next-auth/discussions/8999#discussioncomment-8250545).

 I tried the second approach.

```
import { AuthError } from 'next-auth';

export class CustomAuthError extends AuthError  {
  static type: string;

  constructor(message?: any) {
    super();

    this.type = message;
  }
}
```

```
  async authorize({ email, password }) {
        throw new CustomAuthError('Invalid email or password');
      },
```

Even if I have a try-catch, It won't go to the catch.

```
  try {
      const res = await signIn('credentials', {
        redirect: false,
        username,
        password,
      });
      console.log({ res });
    } catch (error) {
      console.log(JSON.stringify(error));
    }
```

Response is the same

![image](https://github.com/nextauthjs/next-auth/assets/99941065/dcfa7fcb-0233-4fb7-925f-40056472f6ca)

**Reply by SasquatchLV** [reply]
Currently dealing with the same situation, would appreciate if someone could show an example.

**Reply by sourav-dev-prakria** [reply]
am also dealing with the same situation here

**Reply by hieumau12** [reply]
Im also stuck in this situation. But in auth/sveltekit. I thought that it cause by auth/core lib

I found this feat merge from Mar 2
[customizable authorize() error](https://github.com/nextauthjs/next-auth/pull/9871)
but it did not work as expected now

**Comment by jumpjoong** [comment]
```javascript
//Login
const result = await signIn("credentials", {
        // redirect false - 오류 발생시 화면 새로고침 하지 않고 그대로 정지
        redirect: false,
        email: email,
        password: pw,
      });
      if (result?.ok) {
        console.log(result);
      } else {
        console.log(result);
      }

```

```javascript
//POST 

return new Response(
      JSON.stringify({
        error: "Check u r err.",
        ok: false,
      }),
      { status: 401 }
    );

```
```javascript
//[...nextauth]

   async authorize(credentials, req) {
        const res = await fetch(`http://localhost:3000/api/login`, {
          method: "POST",
          headers: {
            "Content-Type": "application/json",
          },
          body: JSON.stringify({
            email: credentials?.email,
            password: credentials?.password,
          }),
        });
        const user = await res.json();
        if (res.ok) {
          return user;
        } else {
          throw new Error(user.error);
        }
        
```
<img width="374" alt="image" src="https://github.com/nextauthjs/next-auth/assets/100519998/98963ba3-01d6-4a90-bc81-557a2852061a">

Is this what you are looking for?

**Reply by dsmabulage** [reply]
> ```js
> {
>         error: "Check u r err.",
>         ok: false,
>       }
> ```

Did you try this approach? I tried it myself and it did not work for me.

```
async authorize({ email, password }) {
        throw new Error(
          {
            error: 'Check u r err.',
            ok: false,
          },
          {
            cause: 'Invalid email or password',
          }
        );
      },
```

![image](https://github.com/nextauthjs/next-auth/assets/99941065/eb19905f-94ec-492c-b2aa-95a47066ad4f)

`"next-auth": "^5.0.0-beta.19"`

**Reply by kimlee3212** [reply]
![image](https://github.com/nextauthjs/next-auth/assets/171894466/c41e44d2-9f75-4450-9617-5921509a603d)


"next-auth": "^4.24.7",

**Reply by kimlee3212** [reply]
![image](https://github.com/nextauthjs/next-auth/assets/171894466/6e9ed060-c318-4f8c-bac8-3677229df0e2)

**Reply by sarowarhosen03** [reply]
> ```js
> //Login
> const result = await signIn("credentials", {
>         // redirect false - 오류 발생시 화면 새로고침 하지 않고 그대로 정지
>         redirect: false,
>         email: email,
>         password: pw,
>       });
>       if (result?.ok) {
>         console.log(result);
>       } else {
>         console.log(result);
>       }
> ```
> 
> ```js
> //POST 
> 
> return new Response(
>       JSON.stringify({
>         error: "Check u r err.",
>         ok: false,
>       }),
>       { status: 401 }
>     );
> ```
> 
> ```js
> //[...nextauth]
> 
>    async authorize(credentials, req) {
>         const res = await fetch(`http://localhost:3000/api/login`, {
>           method: "POST",
>           headers: {
>             "Content-Type": "application/json",
>           },
>           body: JSON.stringify({
>             email: credentials?.email,
>             password: credentials?.password,
>           }),
>         });
>         const user = await res.json();
>         if (res.ok) {
>           return user;
>         } else {
>           throw new Error(user.error);
>         }
>         
> ```
> 
> <img alt="image" width="374" src="https://private-user-images.githubusercontent.com/100519998/342226682-98963ba3-01d6-4a90-bc81-557a2852061a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjAxNjEyMzcsIm5iZiI6MTcyMDE2MDkzNywicGF0aCI6Ii8xMDA1MTk5OTgvMzQyMjI2NjgyLTk4OTYzYmEzLTAxZDYtNGE5MC1iYzgxLTU1N2EyODUyMDYxYS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNzA1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDcwNVQwNjI4NTdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jNGFiZGFhYjNjNWY2NzQwNDEyZjY4M2M4YThkNjQ4OGUwOGU3NzMyMDM3YmU2YWVhYWM1NWJiYTZmMGI3MDFiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.Ha7s5vW33VChhUdm9c0ncXT4pn6ipHfL2SdU8tt0HzE">
> Is this what you are looking for?

In this way Clint side useSession hook data not gonna update

**Comment by Andrew213** [comment]
still issue

**Comment by dileepainivossl** [comment]
still issue

**Comment by hacktolove** [comment]
I HATE NEXT AUTH

**Comment by miracim** [comment]
still issue

**Comment by MasterWolfx** [comment]
stille issue

**Comment by SValab** [comment]
I am experiencing the same thing as well. 

For now my solution is as follows in my signIn callback:

```typescript

Try{
 throw new Error('Invalid sign up session')
} catch (error) {
  if (error instanceof Error) {
    return `/auth/error?error=${error.message}`
  }
}

**Comment by bayunugrohoDev** [comment]
still issue 😠

**Comment by imrim12** [comment]
I have tried to throw the error a several places but the response always retuned successfully, and the response is always a url to the error page.

![image](https://github.com/user-attachments/assets/faffa54b-f8dd-4604-bc64-ca9be90e9cb2)

So i decided to return it as successful from the `authorize` method, then throw it from the `signIn` callback

![image](https://github.com/user-attachments/assets/32dfbf3b-5170-40df-a8fe-9873599f21c9)

This is how i retrieve the error properly 
```
CredentialsProvider({
  authorize() {
    // ...
    return new AuthError(...)
  }
})
```
and in the callbacks
```
callbacks: {
  signIn({ user }) {
    if (user instanceof AuthError) {
      throw user
    }
  }
}
```
In the UI code
```
const response = await signIn('credentials', {
  redirect: false,
  ...credentials.value,
})

if (response.error)
  console.log(response.error) // show your error message here
```

please checkout this answer: https://stackoverflow.com/a/76761731/11762861

**Comment by aaliboyev** [comment]
From the documentation it's not clear enough. After a day of trying different workarounds I understood one thing.

DON'T use signIn function from `"next-auth/react"` in  your NextJs project's client component. This works fine until you come to handle errors. Api methods somehow don't return the code from your custom error class.

I tried the following:

In `@/auth.ts`
```ts
import NextAuth, {CredentialsSignin} from "next-auth"
import Credentials from "next-auth/providers/credentials"
import {login} from "@/services/api/auth";
import {AuthConfig} from "@auth/core";

export class InvalidLoginError extends CredentialsSignin {
    code = 'invalid_credentials'
}

export const authConfig: AuthConfig = {
    providers: [
        Credentials({
            credentials: {
                username: {},
                password: {},
            },
            authorize: async (credentials): Promise<{name: string, accessToken: string}> => {
                const username = credentials.username as string
                const password = credentials.password as string
                try {
                    let loginResponse = await login(username, password)
                    return {
                        name: password,
                        accessToken: loginResponse.data.access_token
                    }
                } catch (e: any) {
                    throw new InvalidLoginError(e.json.detail)
                }
            },
        }),
    ],
}

export const { handlers, signIn, signOut, auth } = NextAuth(authConfig)
```

Remove `signIn ` coming from `"next-auth/react"` and use the one exported from `"@/auth.ts"`
Before directly using it, create `actions.ts` file in your app router, and place the following code inside:
```ts
'use server'
import {InvalidLoginError, signIn} from "@/auth";

export async function authenticate(email: string, password: string) {
    try {
        const r = await signIn("credentials", {
            username: email,
            password: password,
            callbackUrl: "/",
            redirect: false,
        })
        return r
    } catch (error) {
        if (error.cause.err instanceof InvalidLoginError) {
            return {"error": "Incorrect username or password"}
        } else {
            throw new Error("Failed to authenticate")
        }
    }
}
```

Now you are good to use this `authenticate` function in your client component, and it will work perfectly. One issue, as suggested by this discussion, is that it would be better to handle the error directly, skipping `CallbackRouteError`. This would change `error.cause.err instanceof InvalidLoginError` to `error instanceof InvalidLoginError`.

After searching a bit in the source code, I found that there is room for this. Here’s auth.js throwing the exact error if it's an instance of `AuthError`: https://github.com/nextauthjs/next-auth/blob/2adc3a9c2a844d740e5b4366afa80b74fd4c19ed/packages/core/src/lib/actions/callback/index.ts#L524

This means if you extend your custom error from `AuthError`, as @DavidCodina and @Greek suggest:
```ts
export class InvalidLoginError extends AuthError {
    code = 'invalid_credentials' // code is not member of AuthError but does not matter
}
```

You will be able to handle it without looking through `CallbackRouteError`. You can even do:
```ts
export class InvalidLoginError extends AuthError {
    code = 'invalid_credentials'
    constructor(message: string) {
        super(message)
        this.code = message
    }
}
```
and use it for all types of custom errors:
```ts
try {
    // ...
} catch (e: any) {
    throw new InvalidLoginError("invalid_credentials")
}
```

And in server actions, check it:
```ts
try {
    // ...
} catch (e: any) {
    if (e.code === "invalid_credentials") {
        return {error: "Incorrect username or password"}
    } else {
        throw Error("Failed to authenticate")
    }
}
```

**Reply by PineAppleGrits** [reply]
This answer works! 
I seriously can't believe this is documented so bad on the official docs. Thanks!

**Reply by monarcode** [reply]
Worked for me

**Reply by jwenger100** [reply]
Yes this is a head scratcher that such a common use case is buried like this. I am extremely grateful @aaliboyev

**Reply by jhoanborges** [reply]
Worked for me too.

**Reply by nguyenvuhoang** [reply]
But why it redirect set false, set it not return data to login page, i need process something more before i submit to Home.

**Reply by nguyenvuhoang** [reply]
> But why it redirect set false, set it not return data to login page, i need process something more before i submit to Home.

"return r" --> log message "http://localhost:3000/en/login"

**Reply by iamifechi** [reply]
thanks so much for this.

**Comment by Shivendra-Geek** [comment]
still issue

**Comment by mauricabrera** [comment]
For all of you fellas struggling with this, thank God I've managed to find a solution to use on the client side, after some hours! I've based my implementation on this Medium post that point me on the right direction, @aaliboyev response and one hint provided by the new AI on the Auth.js website (ChatGPT doesn't help at all!), the Medium post wasn't working for me using the "next-auth": "^5.0.0-beta.25" version, because it throws an "Access Denied" error with the custom message but inside a "cause" key that I wasn't able to access through the code. Instead what I've done is this, create a new InvalidLoginError class that extends CredentialsSignIn, use signIn callback to change the error message with the custom one we are sending on authorize, read it on the signIn function (I'm using credentials approach) and don't forget to add the 'error' field to the next-auth.d.ts file as follows:

auth.ts

```
import NextAuth, { CredentialsSignin } from "next-auth"
import Credentials from "next-auth/providers/credentials"
import axios from "axios"

class InvalidLoginError extends CredentialsSignin {
  code = "Invalid identifier or password"
  constructor(message: string) {
      super(message)
      this.code = message
  }
}

export const { handlers, signIn, signOut, auth } = NextAuth({
  providers: [
    Credentials({
      name: "Credentials",
      id: "credentials",
      credentials: {
        email: { label: "Email", type: "text" },
        password: { label: "Password", type: "password" },
      },
      async authorize(credentials) {
        try {
          const { data } = await axios.post('customendpoint', {
            email: credentials.email,
            password: credentials.password
          });

          if (!data.status){
            return { error: data.errors }
          } 
          
          return {
            id: data.data.id as string,
            email: data.data.email as string,
            full_name: data.data.full_name,
            role: data.data.role
          }
          
        } catch (error) {
          return { error: error }
        }
      }
    }),
  ],
  session: {
    strategy: "jwt",
    maxAge: (60*60)-7
  },
  callbacks: {
    authorized: async ({ auth }) => {
      return !!auth
    },
    async jwt({ token, user, account, trigger, session }) {
        if (account && user) {
            token.id = user.id
            token.full_name = user.full_name
            token.email = user.email
            token.role = user.role
        }
        return token
    },
    async session({ session, token }) {
        return {
            ...session,
            user: {
                ...session.user,
                id: token.id as string,
                full_name: token.full_name as string,
                email: token.email as string,
                role: token.role as string,
            }
        }
    },
    async signIn({ user }) {
      if (user?.error) {
        throw new InvalidLoginError(user.error)
      }
      return true;
    },
  },
  pages: {
    signIn: '/auth/login'
  },
  trustHost: true
})
```

Function on your page (I'm using formik and yup)

YourLoginScreen.tsx

```
const handleAuth = async () => {
      setLoading(true);
      setError("");
      
      const result = await signIn("credentials", {
        email: formik.values.email,
        password: formik.values.password,
        redirect: false,
      });
      
      if (!result?.ok || result?.error) {
        setError(result?.code ? result.code : "")
        setLoading(false)
      } else {
        router.push('/dashboard');
      }      
  };

  const formik = useFormik({
    initialValues: {
      email: "",
      password: "",
    },
    validationSchema: Yup.object().shape({
      email: Yup.string().email('*Enter a valid e-mail').required('*Required'),
      password: Yup.string().required('*Required'),
    }),
    onSubmit: (values) => {
      handleAuth();
    },
  });
```
  
next-auth.d.ts

  ```
import { DefaultSession, DefaultUser } from "next-auth";

export enum Role {
  admin = "ADMIN",
  cashier = "CASHIER",
}
declare module "next-auth" {
  interface User {
    id?: number;
    full_name?: string;
    role?: Role | string;
    error? : string;
  }

  interface Session {
      user: User & DefaultSession["user"]
      expires: string
  }
}
```
  
Hope it helps someone, have an amazing day!

**Reply by Youngpro95** [reply]
It works!! thanks

**Reply by ducviet1696** [reply]
Why don't you throw InvalidLoginError directly in the authorize function, but instead return a User with an error field and then finally throw it in signIn callback? This forces you to make the entire state of User optional, and that seems like a maze of mistakes

**Comment by yashdilip** [comment]
I have my setup as

```
@/auth.ts

class InvalidLoginError extends CredentialsSignin {
  code = "custom_error"

  constructor(message: string) {
    super(message)
    this.code = message
  }
}
```

Inside `authorize()` used as

```if (user.status !== "active") throw new InvalidLoginError("UserInactive");```

and In client side I have error map

```
const errorMap: Record<string, string> = {
  UserInactive: "Your account is currently inactive.",
  // Add other custom errors here
}


if (result?.error || result?.code) {
        // The result.error here will be the *generic* error name (e.g., "MfaRequired")
        const errorMessage = errorMap[result?.code as string] || errorMap.Configuration;
        setError(errorMessage); // Set the user-friendly message in your component state
        setIsLoading(false);

        return;
}
```

Thanks to @mauricabrera and @ducviet1696
